pkg/util/resource/monitor.go

```diff
@@ -0,0 +1,225 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"runtime/metrics"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/procfs"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/configs"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type ExhaustedError struct{}
+
+func (e *ExhaustedError) Error() string {
+	return "resource exhausted"
+}
+
+const heapMetricName = "/memory/classes/Heap/objects:bytes"
+const monitorInterval = time.Second
+const dataPointsToAvg = 30
+
+type IScanner interface {
+	Scan() (Stats, error)
+}
+
+type Scanner struct {
+	proc          procfs.Proc
+	metricSamples []metrics.Sample
+}
+
+type Stats struct {
+	CPU  float64
+	Heap uint64
+}
+
+func NewScanner() (*Scanner, error) {
+	proc, err := procfs.Self()
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading proc directory")
+	}
+
+	metricSamples := make([]metrics.Sample, 1)
+	metricSamples[0].Name = heapMetricName
+	metrics.Read(metricSamples)
+
+	for _, sample := range metricSamples {
+		if sample.Value.Kind() == metrics.KindBad {
+			return nil, fmt.Errorf("metric %s is not supported", sample.Name)
+		}
+	}
+
+	return &Scanner{
+		proc:          proc,
+		metricSamples: metricSamples,
+	}, nil
+}
+
+func (s *Scanner) Scan() (Stats, error) {
+	stat, err := s.proc.Stat()
+	if err != nil {
+		return Stats{}, err
+	}
+
+	metrics.Read(s.metricSamples)
+
+	return Stats{
+		CPU:  stat.CPUTime(),
+		Heap: s.metricSamples[0].Value.Uint64(),
+	}, nil
+}
+
+type Monitor struct {
+	services.Service
+
+	scanner IScanner
+
+	containerLimit configs.Resources
+	utilization    configs.Resources
+	thresholds     configs.Resources
+
+	// Variables to calculate average CPU utilization
+	index         int
+	cpuRates      []float64
```

Comment:
Maybe it makes sense to use an array instead of a slice here? `dataPointsToAvg` is a constant either way.

Reply:
True, will change it